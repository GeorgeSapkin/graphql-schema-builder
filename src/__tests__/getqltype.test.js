'use strict';

const graphql = require('graphql');

const {
  GraphQLDateTime
} = require('graphql-iso-date');

const {
  types: {
    Mixed,
    ObjectId
  }
} = require('../types');

const {
  getQLType: _getQLType
} = require('../getqltype');

const {
  Asset,
  nop,
  schemaStore
} = require('./__fixtures__');

const {
  GraphQLNonNull,
  GraphQLObjectType
} = graphql;

const getQLType = _getQLType(graphql);

describe('getQLType', () => {
  describe('returns', () => {
    it('String!', () => {
      expect(
        getQLType(nop, { type: String })
      ).toMatchSnapshot();
    });

    it('String', () => {
      expect(
        getQLType(nop, { type: String, required: false })
      ).toMatchSnapshot();
    });

    it('Float!', () => {
      expect(
        getQLType(nop, { type: Number })
      ).toMatchSnapshot();
    });

    it('Float', () => {
      expect(
        getQLType(nop, { type: Number, required: false })
      ).toMatchSnapshot();
    });

    it('DateTime!', () => {
      expect(
        getQLType(nop, { type: Date })
      ).toMatchObject(
        new GraphQLNonNull(GraphQLDateTime)
      );
    });

    it('DateTime', () => {
      expect(
        getQLType(nop, { type: Date, required: false })
      ).toMatchSnapshot();
    });

    it('Boolean!', () => {
      expect(
        getQLType(nop, { type: Boolean })
      ).toMatchSnapshot();
    });

    it('Boolean', () => {
      expect(
        getQLType(nop, { type: Boolean, required: false })
      ).toMatchSnapshot();
    });

    it('JSON!', () => {
      expect(
        getQLType(nop, { type: Mixed })
      ).toMatchSnapshot();
    });

    it('JSON', () => {
      expect(
        getQLType(nop, { type: Mixed, required: false })
      ).toMatchSnapshot();
    });

    it('ID!', () => {
      expect(
        getQLType(nop, { type: ObjectId })
      ).toMatchSnapshot();
    });

    it('ID', () => {
      expect(
        getQLType(nop, { type: ObjectId, required: false })
      ).toMatchSnapshot();
    });

    it('refType!', () => {
      const getExistingType = () => new GraphQLObjectType(Asset);

      expect(
        getQLType(getExistingType, {
          type: ObjectId,
          ref:  Asset.name
        })
      ).toMatchSnapshot();
    });

    it('refType', () => {
      const getExistingType = schemaStore.get.bind(schemaStore);

      expect(
        getQLType(getExistingType, {
          type: ObjectId, ref: Asset.name, required: false
        })
      ).toMatchSnapshot();
    });

    it('stringType', () => {
      const getExistingType = schemaStore.get.bind(schemaStore);

      expect(
        getQLType(getExistingType, {
          type: 'Asset', required: false
        })
      ).toMatchSnapshot();
    });

    it('stringType!', () => {
      const getExistingType = schemaStore.get.bind(schemaStore);

      expect(
        getQLType(getExistingType, {
          type: 'Asset'
        })
      ).toMatchSnapshot();
    });

    it('ID!', () => {
      expect(
        getQLType(nop, { type: ObjectId })
      ).toMatchSnapshot();
    });

    it('[String]!', () => {
      expect(
        getQLType(nop, {
          type:     [{ type: String }],
          required: false
        })
      ).toMatchSnapshot();
    });

    it('[Float!]', () => {
      expect(
        getQLType(nop, { type: [Number], required: false })
      ).toMatchSnapshot();
    });

    it('[Boolean]!', () => {
      expect(
        getQLType(nop, { type: [{ type: Boolean, required: false }] })
      ).toMatchSnapshot();
    });

    it('[refType!]!', () => {
      const getExistingType = schemaStore.get.bind(schemaStore);

      expect(
        getQLType(getExistingType, {
          type: [ObjectId],
          ref:  Asset.name
        })
      ).toMatchSnapshot();
    });

    it('[stringType!]!', () => {
      const getExistingType = schemaStore.get.bind(schemaStore);

      expect(
        getQLType(getExistingType, {
          type: ['Asset']
        })
      ).toMatchSnapshot();
    });

    it('null', () => expect(getQLType(nop, { type: {} })).toMatchSnapshot());
  });

  describe('throws', () => {
    it('without schemaStore', () => expect(getQLType).toThrow());

    it('with bad schemaStore', () => expect(() => getQLType({})).toThrow());

    it('without type', () => expect(() => getQLType(new Map())).toThrow());
  });
});
