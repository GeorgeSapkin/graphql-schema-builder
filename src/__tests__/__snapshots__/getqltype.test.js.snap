// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getQLType returns [Boolean]! 1`] = `"[Boolean]!"`;

exports[`getQLType returns [Float!] 1`] = `"[Float!]"`;

exports[`getQLType returns [String]! 1`] = `"[String!]"`;

exports[`getQLType returns [refType!]! 1`] = `"[Asset!]!"`;

exports[`getQLType returns [stringType!]! 1`] = `"[Asset!]!"`;

exports[`getQLType returns Boolean 1`] = `"Boolean"`;

exports[`getQLType returns Boolean! 1`] = `"Boolean!"`;

exports[`getQLType returns DateTime 1`] = `"DateTime"`;

exports[`getQLType returns Float 1`] = `"Float"`;

exports[`getQLType returns Float! 1`] = `"Float!"`;

exports[`getQLType returns ID 1`] = `"ID"`;

exports[`getQLType returns ID! 1`] = `"ID!"`;

exports[`getQLType returns ID! 2`] = `"ID!"`;

exports[`getQLType returns JSON 1`] = `"JSON"`;

exports[`getQLType returns JSON! 1`] = `"JSON!"`;

exports[`getQLType returns String 1`] = `"String"`;

exports[`getQLType returns String! 1`] = `"String!"`;

exports[`getQLType returns null 1`] = `null`;

exports[`getQLType returns refType 1`] = `"Asset"`;

exports[`getQLType returns refType! 1`] = `"Asset!"`;

exports[`getQLType returns stringType 1`] = `"Asset"`;

exports[`getQLType returns stringType! 1`] = `"Asset!"`;
